Add mesh shading info to naga IR

inner-daemons opened the pull request:
**Connections**
Mostly split off from #7930 
Works towards #7197 

**Description**
This PR adds mesh shading info to naga IR so that parsers and writers have an interface to use.

**Testing**
No testing yet (coming in later PRs, the code here has been tested in #7930)

**Squash or Rebase?**

Squash

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

inner-daemons committed:
Initial commit

inner-daemons committed:
Other initial changes

inner-daemons committed:
Updated shader snapshots

inner-daemons commented:
I'm unsure why the MSRV minimal versions thing is failing, it doesn't look related to this PR since it happens in another crate and this PR doesn't touch anything cargo.

inner-daemons committed:
Added new HLSL limitation

inner-daemons committed:
Moved error to global variable error

cwfitzgerald commented:
Should be fixed with #8112, thanks @andyleiserson!

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy commented on naga/src/ir/mod.rs:
```diff
@@ -966,6 +984,7 @@ pub enum Binding {
         /// Optional `blend_src` index used for dual source blending.
         /// See <https://www.w3.org/TR/WGSL/#attribute-blend_src>
         blend_src: Option<u32>,
+        per_primitive: bool,
```
Could we have some docs for this? The `Location` variant's docs talk about passing values from the vertex stage to the fragment stage; we should make sure the story told here makes sense for readers working on mesh shaders too.

inner-daemons replied:
I've tried to write some docs. Let me know if you have further comments.

inner-daemons committed:
Added docs to per_primitive

inner-daemons committed:
Added a little bit more docs here and there in IR

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Adding validation to ensure that task shaders have a task payload

inner-daemons committed:
Updated spec to reflect the change to payload variables

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons commented:
Going to ping you again @jimblandy just to make sure that this doesn't get forgotten about until the next meeting :)

jimblandy commented:
@SupaMaggie70Incorporated I'll try to get to this tonight, but it may be Saturday.

jimblandy commented:
@SupaMaggie70Incorporated Are these comments at the top of `mesh_shading.md` still accurate?

> Currently `naga` has no support for mesh shaders beyond recognizing the additional shader stages. For this reason, all shaders must be created with `Device::create_shader_module_passthrough`.

inner-daemons committed:
Updated the mesh shading spec because it was goofy

inner-daemons commented:
> @SupaMaggie70Incorporated Are these comments at the top of `mesh_shading.md` still accurate?
> 
> > Currently `naga` has no support for mesh shaders beyond recognizing the additional shader stages. For this reason, all shaders must be created with `Device::create_shader_module_passthrough`.

I've updated the top of `mesh_shading.md`. Until further PRs it remains true that you will need to use passthrough shaders but it now has more "processing" ability so the statement wasn't strictly true.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy commented:
(This is not a review comment on this PR, and not directed at SupaMaggie, just commenting on the pre-existing state of the docs.)

> ***This is not*** a thorough explanation of mesh shading and how it works. Those wishing to understand mesh shading more broadly should look elsewhere first.

This is just not adequate. If we're going to add syntax to Naga that is not covered by the WGSL specification, we need to document that syntax. Knowing how mesh shading works in Vulkan does not magically explain the syntax for mesh shaders in WGSL, or how to invoke them in wgpu.

inner-daemons commented:
> (This is not a review comment on this PR, and not directed at SupaMaggie, just commenting on the pre-existing state of the docs.)
> 
> > _**This is not**_ a thorough explanation of mesh shading and how it works. Those wishing to understand mesh shading more broadly should look elsewhere first.
> 
> This is just not adequate. If we're going to add syntax to Naga that is not covered by the WGSL specification, we need to document that syntax. Knowing how mesh shading works in Vulkan does not magically explain the syntax for mesh shaders in WGSL, or how to invoke them in wgpu.

It wouldn't be part of this PR, but if you think that having a writeup somewhere to describe mesh shaders would be useful, I'm happy to do that separately. However, I have never written a comprehensive GPU API spec before, so I don't exactly think it would be of comparable quality to e.g. the webgpu spec :)

Also, that specific disclaimer was copy-pasted directly from the RT spec.

jimblandy committed:
some doc tweaks

jimblandy commented:
> It wouldn't be part of this PR, but if you think that having a writeup somewhere to describe mesh shaders would be useful, I'm happy to do that separately. However, I have never written a comprehensive GPU API spec before, so I don't exactly think it would be of comparable quality to e.g. the webgpu spec :)

Well, the alternative is just saying "read the code and the examples and figure it out". You want people to actually *use* your work, right?

I just pushed some docs; do they look correct?

jimblandy commented:
More later - sorry to take so long.

inner-daemons commented:
> > It wouldn't be part of this PR, but if you think that having a writeup somewhere to describe mesh shaders would be useful, I'm happy to do that separately. However, I have never written a comprehensive GPU API spec before, so I don't exactly think it would be of comparable quality to e.g. the webgpu spec :)
> 
> Well, the alternative is just saying "read the code and the examples and figure it out". You want people to actually _use_ your work, right?
> 
> I just pushed some docs; do they look correct?

I'd like to make a few edits, at least around the section about task shader returns, just to clarify some things. Overall I think this is a step in the right direction. I'll try to get to this tonight or tomorrow

jimblandy commented:
We don't need to have everything fully documented before we land. I just end up writing docs (or trying to) in the process of figuring out what's going on in a PR.

jimblandy commented:
> Also, that specific disclaimer was copy-pasted directly from the RT spec.

Fair.

jimblandy commented:
> I'd like to make a few edits, at least around the section about task shader returns, just to clarify some things. Overall I think this is a step in the right direction. I'll try to get to this tonight or tomorrow

Sounds great. Just go ahead and push your changes to the branch, and I'll check them out. I'm curious to see what changes, but I'm not possessive about the text.

inner-daemons commented:
@jimblandy You mention invocations a lot in your docs, is this a mistake or is this how wgpu docs are commonly written? Task and mesh shaders all function in units of workgroups (like compute shaders) and all outputs are shared across the workgroups.

The workgroup size is also constant and part of the shader. The group size arguments you pass to `draw_mesh_tasks` or that a task shader outputs correspond to the number of workgroups and not the size of a single workgroup.

inner-daemons committed:
Tried to clarify docs a little

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons commented:
I'm going to work on the spec a little more to try to clarify what mesh shaders *are*, not just how the new API works.

inner-daemons committed:
Tried to update spec

inner-daemons committed:
Removed a warning

jimblandy commented:
> @jimblandy You mention invocations a lot in your docs, is this a mistake or is this how wgpu docs are commonly written?

My mistake.

> The workgroup size is also constant and part of the shader. The group size arguments you pass to `draw_mesh_tasks` or that a task shader outputs correspond to the number of workgroups and not the size of a single workgroup.

Mistake. This one I should have known - thanks for catching it.

jimblandy commented on wgpu/src/api/render_pass.rs:
```diff
@@ -226,7 +226,33 @@ impl RenderPass<'_> {
+    ///
+    /// The current pipeline must be a mesh shader pipeline.
+    ///
+    /// If the current pipeline has a task shader, run it with an workgroup for
+    /// every `vec3<u32>(i, j, k)` where `i`, `j`, and `k` are between `0` and
+    /// `group_count_x`, `group_count_y`, and `group_count_z`. The invocation with
+    /// index zero in each group is responsible for determining the mesh shader dispatch.
+    /// Its return value indicates the number of workgroups of mesh shaders to invoke. It also
+    /// passes a payload value for them to consume. Because each task workgroup is essentially
+    /// a mesh shader draw call, mesh workgroups dispatched by different task workgroups
+    /// cannot interact in any way, and `workgroup_id` corresponds to its location in the
+    /// calling specific task shader's dispatch group.
```
This is great detail. Thanks very much for the explanation.

I do wonder if there's a better place for this, but that can come later.

jimblandy commented on wgpu/src/api/render_pass.rs:
```diff
@@ -226,7 +226,33 @@ impl RenderPass<'_> {
+    /// there is no global vertex buffer. These primitives are passed to the rasterizer and
+    /// essentially treated like a vertex shader output, except that the mesh shader may
+    /// choose to cull specific primitives or pass per-primitive non-interpolated values
+    /// to the mesh shader. As such, each primitive is then rendered with the current
```
Should this be "to the fragment shader"?

inner-daemons replied:
Updated

inner-daemons committed:
Addressed comment about docs mistake

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy committed:
Review in progress

- Extensive revisions to `docs/api-specs/mesh_shading.md`.

- Doc comments.

- Ensure `Module` stays at the bottom of `ir/mod.rs`.

- Avoid a clone.

- Rename some arguments to be more specific.

- Minor readability tweaks.

inner-daemons commented:
@jimblandy Not to bug you *too* much but this is really the one PR that I want to get in as soon as possible, since it blocks everything else (which will be harder to work on as school ramps up). So if you could prioritize getting a review and subsequent merge like within the next week that'd be very much appreciated.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons commented:
@jimblandy This PR has now been dead for more than 3 weeks, will you have time to review it before the face to face?

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons commented:
@jimblandy I see your comment in the meeting notes about getting to it today or reassigning it, I will hold you to that :)

At this point the only deadline is the next release (28.0) so I'm not in a huge rush, more just hoping that subsequent PRs that will likely require more back-and-forth can be landed in a more reasonable amount of time.

ErichDonGubler commented:
@SupaMaggie70Incorporated: But...our next release is planned to be today, innit? Or were you referring to your own downstream release?

inner-daemons commented:
> @SupaMaggie70Incorporated: But...our next release is planned to be today, innit? Or were you referring to your own downstream release?

@ErichDonGubler I meant the release after this one. My goal had been for this release but that (obviously) won't be happening.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy committed:
mesh_shading.md: more tweaks

jimblandy commented:
I was finally able to make time to look at this again today. I've pushed a commit with some further changes; please check them out to see if they look right. Here's my current review checklist for the code:

- Reviewing Naga mesh shader patches
  - [ ] `enable mesh_shading` required?
  - [ ] @payload checks
    - [ ] is @payload optional on @task shaders, or required? It's optional in SPIR-V.
    - [ ] size/type restrictions?
  - [ ] @task checks
    - [ ] @workgroup size attribute present
    - [ ] Returns vec3<u32>? 
    - [ ] Return type marked with `@builtin(mesh_task_size)`?
  - [ ] SPIR-V requirements
    - [ ] TaskEXT
    - [ ] MeshEXT
    - [ ] OutputPrimitivesEXT
    - [ ] OutputVertices
  - [ ] @mesh checks
    - [ ] @workgroup size attribute present
    - [ ] has no return value
    - [ ] has a @primitive_output(P, PN) attribute
      - [ ] P is a struct type
      - [ ] Exactly one of the following @builtins is present: triangle_indices, line_indices, point_indices
      - [ ] @builtin / @locations are unique (perhaps already enforced)
      - [ ] Every member with @location must also have @per_primitive
      - [ ] conditions on PN?
    - [ ] has a @vertex_output(V, VN) attribute
      - [ ] V satisfies the same requirements as a @vertex shader return value
  - [ ] @vertex checks
    - [ ] no @per_fragment on @location outputs
  - [ ] @fragment checks
    - [ ] @per_primitive attribute is present only on inputs that have @location
  - [ ] pipeline creation time checks
    - [ ] @task -> @mesh
      - [ ] if a @task shader is supplied, and has @payload, the @mesh shader must have the same @payload
      - [ ] if a @task shader is not supplied, or lacks @payload, the @mesh shader must not have @payload
    - [ ] @mesh -> @fragment (should be the same as existing @vertex -> @fragment, plus @per_primitive checks)
      - [ ] @mesh @location outputs are one-to-one with @fragment @location inputs
      - [ ] @mesh @location outputs are @per_primitive iff the corresponding @fragment @location input is too
  - [ ] General
    - [ ] If your change iterates over a collection, did you ensure the
          order of iteration was deterministic? Using `HashMap` and
          `HashSet` is fine, as long as you don't iterate over it.
    - [ ] If you insert elements into a set or map that you expect are not
          already present, did you make an assertion about `insert`'s
          return value?
    - [ ] If you added a new feature to WGSL that is not covered by the
          WebGPU specification:
      - [ ] Did you add a `Capability` flag for it?
      - [ ] Did you document the feature fully in that flag's doc comment?
      - [ ] Did you ensure the validator rejects programs that use the
            feature unless its capability is enabled?
    - If your change adds or removes `Handle`s from the IR:
      - [ ] Did you update handle validation in `valid::handles`?
      - [ ] Did you update the compactor in `compact`?
      - [ ] Did you update `back::pipeline_constants::adjust_expr`?
    - If your change adds a new operation:
      - [ ] Did you update the typifier in `proc::typifier`?
      - [ ] Did you update the validator in `valid::expression`?
      - [ ] If the operation can be used in constant expressions, did you
            update the constant evaluator in `proc::constant_evaluator`?
    - [ ] If your change introduces any new identifiers to generated code,
          how did you ensure they won't conflict with the users'
          identifiers? (This is usually not relevant to the SPIR-V
          backend.)
      - [ ] Did you use the `Namer` to generate a fresh identifier?
      - [ ] Did you register the identifier as a reserved word with the the `Namer`?
      - [ ] Did you use a reserved prefix registered with the `Namer`?

inner-daemons commented on docs/api-specs/mesh_shading.md:
```diff
@@ -151,19 +151,19 @@ A function with the `@mesh` attribute is a **mesh shader entry point**. Mesh sha
 - `@primitive_output(P, NP)`: This indicates that the mesh shader workgroup will generate at most `NP` primitives, each of type `P`.
 
-Each mesh shader entry point invocation must call the `setMeshOutputs(numVertices: u32, numPrimitives: u32)` builtin function exactly once, in uniform control flow. The values passed by each workgroup's first invocation (that is, the one whose `local_invocation_index` is `0`) determine how many vertices (values of type `V`) and primitives (values of type `P`) the workgroup must produce. This call ...
+Before generating any results, each mesh shader entry point invocation must call the `setMeshOutputs(numVertices: u32, numPrimitives: u32)` builtin function exactly once, in uniform control flow. The values passed by each workgroup's first invocation (that is, the one whose `local_invocation_index` is `0`) determine how many vertices (values of type `V`) and primitives (values of type `P`) the wo ...
```
I think it used to be this, but after writing the SPIR-V writer I decided that would be unnecessary. On SPIR-V, we have a temporary output buffer that is then copied from when finishing execution, so the order doesn't matter. I think we would do the same with HLSL. I haven't thought about MSL but that should work fine too.

inner-daemons replied:
It might still good to document this limitation, even if it isn't a concern yet, as it would also allow us to change writer behavior down the line.

SupaMaggie70Incorporated commented on docs/api-specs/mesh_shading.md:
```diff
@@ -151,19 +151,19 @@ A function with the `@mesh` attribute is a **mesh shader entry point**. Mesh sha
 
 A mesh shader entry point must have the following attributes:
 
+- `@workgroup_size`: this has the same meaning as when it appears on a compute shader entry point.
```
I'll make a similar change to the same little bit of text for task shaders

inner-daemons commented:
@jimblandy Thanks for getting to this! The code/documentation changes all look great to me. I'll probably spend some time looking over the spec since it looks like you focused a lot on that. The task list you laid out looks long and intimidating but I think each individual item should be very quick. Hopefully we can get this in soon-ish! And hopefully that represents a more solidified API that won't be changing much as I realize more limitations.

One note though, some of the items aren't covered in this PR. The SPIR-V changes come later, and I really want to push pipeline creation checks to another PR so I can get this one in ASAP and have multiple PRs in the works afterward. The last few points though definitely I will think about on my WGSL parser and SPIR-V writer PRs

inner-daemons committed:
Ran cargo fmt

inner-daemons commented on docs/api-specs/mesh_shading.md:
When #8206 lands, this will change to `max_mesh_multiview_view_count` I think. Something to keep in mind, for whichever lands first (hopefully this one!)

inner-daemons commented on docs/api-specs/mesh_shading.md:
```diff
@@ -99,34 +133,57 @@ Using any of these features in a `wgsl` program will require adding the `enable
+
+A mesh shader entry point must have a `@workgroup_size` attribute, meeting the same requirements as one appearing on a compute shader entry point.
+
+If the mesh shader pipeline has a task shader entry point with a `@payload(G)` attribute, then the pipeline's mesh shader entry point must also have a `@payload(G)` attribute, naming the same variable. Mesh shader invocations can read, but not write, this variable, which is initialized to whatever value was written to it by the task shader workgroup that dispatched this mesh shader grid.
```
HLSL requires that task shaders have a non-zero sized task payload, so we will probably reflect that in naga.

inner-daemons commented on docs/api-specs/mesh_shading.md:
```diff
@@ -99,34 +133,57 @@ Using any of these features in a `wgsl` program will require adding the `enable
+- `cull_primitive`: The annotated member must be of type `bool`. If it is true, then the primitive is skipped during rendering.
 
-Additionally, the `@location` attributes from the vertex and primitive outputs can't overlap.
+Every member of `P` with a `@location` attribute must either have a `@per_primitive` attribute, or be part of a struct type that appears in the primitive data as a struct member with the `@per_primitive` attribute.
```
I'm not exactly sure what this means. Maybe we should just require that the `@per_primitive` attribute only be applied to struct members with `@location`. Unsure

inner-daemons commented:
I'm super super pleased in general with the new spec :) No further comments

Just one thing I want to note. This is the naga IR pull request, if we could defer most of the spec/documentation work to later PRs and tweaks I'd appreciate it since as I mentioned I kinda want the code here to get in sooner rather than later.

inner-daemons committed:
Small tweaks

jimblandy commented:
Yes, hopefully the checklist is fine-grained enough that it won't actually take that long.

I see that the backends are stubbed out for now, but I wanted to at least write down all the validation that seemed to be necessary. If some of the checklist items are carried over to later PRs that's fine.

jimblandy commented:
> I think it used to be this, but after writing the SPIR-V writer I decided that would be unnecessary. On SPIR-V, we have a temporary output buffer that is then copied from when finishing execution, so the order doesn't matter. I think we would do the same with HLSL. I haven't thought about MSL but that should work fine too.

@SupaMaggie70Incorporated So the idea is that you wouldn't do the `OpSetMeshOutputsEXT` until the very end of the function, and it'd be followed by a copy? Because the [SPV_EXT_mesh_shader spec][spec] says: 

> `OpSetMeshOutputsEXT` must be called before any variable from `Output` storage class is written to.

[spec]: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_primitive_instructions

That would require two copies of the vertex and primitive buffers. Is there no way to avoid that?

inner-daemons commented:
> > I think it used to be this, but after writing the SPIR-V writer I decided that would be unnecessary. On SPIR-V, we have a temporary output buffer that is then copied from when finishing execution, so the order doesn't matter. I think we would do the same with HLSL. I haven't thought about MSL but that should work fine too.
> 
> @SupaMaggie70Incorporated So the idea is that you wouldn't do the `OpSetMeshOutputsEXT` until the very end of the function, and it'd be followed by a copy? Because the [SPV_EXT_mesh_shader spec](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_primitive_instructions) says:
> 
> > `OpSetMeshOutputsEXT` must be called before any variable from `Output` storage class is written to.
> 
> That would require two copies of the vertex and primitive buffers. Is there no way to avoid that?

In the SPIR-V writer part I document why there *must* be a temporary buffer for both. It boils down to the fact that the output buffers must match the max vertices and max primitives for the entry point. So if a function is called by multiple mesh shader entry points with differnet max vertex and max primitive counts, it can still only write to one buffer, so a private variable with size equal to the largest of the entry points is used. Its super hacky and unfortunate, but that's the best I could come up with.

The idea would be that you could call `setMeshOutputs` anywhere, multiple times, and possibly not at all (it would default to zero). Then when the entry point exits, it would call that and then write all of the vertices/primitives. The other benefit of this approach is that validation is much easier, since when exiting you never write past the values given in setMeshOutputs without giving the user the opportunity to write beyond the limit.

inner-daemons commented:
I'm gonna stop haggling you about this PR, but I'd like to say again that if this is the process we use for the following PRs, wgpu will never support mesh shading.

Edit: I no longer stand by the first statement

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons commented:
@jimblandy I retract previous promises of stopping the harassment. Also,
<img width="564" height="521" alt="image" src="https://github.com/user-attachments/assets/b9c3aeb2-960d-4b70-8847-64130320cb77" />

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy committed:
[naga] Move definition of `ShaderStage::compute_like` to `proc`.

Move the definition of `naga::ShaderStage::compute_like` from
`naga::ir` into `naga::proc`. We generally want ot keep methods out of
`naga::ir`, since the IR itself is complicated enough already.

jimblandy committed:
Replace TODO comment with followup issue.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy commented on naga/src/valid/analyzer.rs:
```diff
@@ -1158,6 +1209,54 @@ impl FunctionInfo {
+        if let &Some(ref other_vertex) = &callee.vertex_type {
+            self.try_update_mesh_vertex_type(other_vertex.0, other_vertex.1)?;
+        }
+        if let &Some(ref other_primitive) = &callee.vertex_type {
```
Shouldn't this be:

```suggestion
        if let &Some(ref other_primitive) = &callee.primitive_type {
```

Let's add a test to `naga/tests/naga/validation.rs` for this, too.

inner-daemons replied:
I'm not sure what such a test would look like

jimblandy committed:
minor changes for readability

inner-daemons committed:
Update analyzer.rs

Co-authored-by: Jim Blandy <jimb@red-bean.com>

jimblandy committed:
Pull mesh shader output type validation out into its own function.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy commented on naga/src/valid/interface.rs:
```diff
@@ -742,11 +851,27 @@ impl super::Validator {
+                && (!result_built_ins.is_empty() || !self.location_mask.is_empty())
+            {
+                return Err(EntryPointError::UnexpectedMeshShaderEntryResult.with_span());
+            }
```
This is surprising: I thought mesh shaders can't return anything. Can't we just raise an error here whenever `stage == Mesh` simply because `ep.function.result` is `Some`?

inner-daemons replied:
We could. I was thinking here about the possibility of mesh shaders somehow in the future returning something globally. For example we could drop the `setMeshOutputs` thing and simply have the vertex and primitive count be part of the return value. Not strictly necessary, probably can be done away with for now. This is already how task shaders function basically.

inner-daemons committed:
Removed stuff in accordance with Jim's recommendation

jimblandy commented on naga/src/valid/interface.rs:
```diff
@@ -683,18 +769,39 @@ impl super::Validator {
             .validate_function(&ep.function, module, mod_info, true)
             .map_err(WithSpan::into_other)?;
 
+        if let Some(handle) = ep.task_payload {
```
It seems like there are actually two copies of the `task_payload` checking in this function, but perhaps I'm misreading.

jimblandy replied:
fixed in 1173b0f

jimblandy committed:
doc fixes

jimblandy committed:
remove duplicated task payload validation

cwfitzgerald commented on naga/src/back/pipeline_constants.rs:
```diff
@@ -296,6 +299,28 @@ fn process_workgroup_size_override(
+            mesh_info.max_primitives = module
+                .to_ctx()
+                .eval_expr_to_u32(adjusted_global_expressions[r#override])
+                .map_err(|_| PipelineConstantError::NegativeWorkgroupSize)?;
```
This probably should be a different error message.

cwfitzgerald commented on naga/src/proc/mod.rs:
```diff
@@ -628,6 +631,16 @@ pub fn flatten_compose<'arenas>(
 }
 
+impl super::ShaderStage {
+    // TODO: make more things respect this
```
Could you file an issue about this and link it here?

inner-daemons replied:
I think a while back I did a naive search by reference for `ShaderStage::Compute` and tried to change everything that seemed relevant. I'll give it another look but we might not have more work to do

inner-daemons replied:
Yeah it seems like most of the related changes here will come later. E.g. when the frontends or backends get added. I'll do an initial commit but this should be considered part of whichever PR could expose the lack of awareness of compute-like shaders

cwfitzgerald commented on naga/src/valid/analyzer.rs:
```diff
@@ -1113,6 +1147,34 @@ impl FunctionInfo {
+                    } => {
+                        let _ = self.add_ref(vertex_count);
+                        let _ = self.add_ref(primitive_count);
+                        FunctionUniformity::new()
```
I'm fairly certain that this is wrong unless we're going to guarentee that on all backends we will internally buffer it and then flush afterwards. I expect that these need to be called in uniform control flow, though I would need to check specs on this.

inner-daemons replied:
That's indeed the intention. I explain why in certain parts of the SPIR-V writer, but it boils down to the fact that

* When writing to the output array in any language, the output array must be sized exactly correctly
* Certain other functions that might be used by multiple entry points might want to write to each of those entry points, but without locking itself into one array size (if the entry points have same mesh types but different max counts)
* Therefore, they must write to a temporary array that gets copied

Also, this will make validation easier since we prevent writing past specified count automatically and we don't have to check that `setMeshOutputs` always gets called exactly once before any of the `setVertex` or `setPrimitive`'s get called

inner-daemons replied:
I think uniform control flow isn't necessary for `setVertex` and `setPrimitive` since each thread can write its own little piece of memory and I don't think writing to memory needs to be uniform. I believe `setMeshOutputs` currently functions so that only the first thread actually writes it so in that case uniform control flow is probably more necessary, since if threads `1..n` instead of `0..n` write or something then nothing would actually be written. Maybe we should change that behavior though, e.g. use workgroup atomics and advise the user to only call it on one thread

cwfitzgerald replied:
This was discussed and because of buffering this isn't a problem.

cwfitzgerald commented on naga/src/valid/function.rs:
```diff
@@ -217,6 +217,14 @@ pub enum FunctionError {
     #[error("Expression not visited by the appropriate statement")]
     UnvisitedExpression(Handle<crate::Expression>),
+    #[error("Expression {0:?} should be u32, but isn't")]
+    InvalidMeshFunctionCall(Handle<crate::Expression>),
```
Could you make this error more expressive, as a user, this doesn't give me much context to go off of, especially if the expression is displayed as `[124]`

inner-daemons replied:
Just tried to address this, I'll leave it open since I'm not great at writing error messages

cwfitzgerald commented on naga/src/valid/function.rs:
```diff
@@ -1539,6 +1547,40 @@ impl super::Validator {
+                        crate::MeshFunction::SetVertex { index, value: _ }
+                        | crate::MeshFunction::SetPrimitive { index, value: _ } => {
+                            ensure_u32(index)?;
+                            // TODO: ensure it is correct for the value
```
What's this about? Will it be handled in a future PR, or should this be an issue?

inner-daemons replied:
Err, this should be fine. I think that type checking for the value is handled elsewhere

cwfitzgerald commented on naga/src/valid/interface.rs:
```diff
@@ -327,6 +394,24 @@ impl VaryingContext<'_> {
+                    // shader.
+                } else {
+                    // All other `Location` bindings must not be `per_primitive`.
+                    if per_primitive {
```
Nit - else if

inner-daemons replied:
Surprised clippy missed this

cwfitzgerald commented on naga/src/valid/interface.rs:
```diff
@@ -390,11 +475,12 @@ impl VaryingContext<'_> {
                     }
                 }
 
+                // TODO: update this to reflect the fact that per-primitive outputs aren't interpolated for fragment and mesh stages
```
Does this need to be dealt with?

cwfitzgerald reviewed (changes requested):
Alright, with knowledge that front/backend support and wgpu integration is coming later, and tests will come with that. Some comments, particularly about TODOs, but nothing major.

inner-daemons committed:
Quick little changes

inner-daemons committed:
Another quick fix

jimblandy committed:
doc fixes

inner-daemons committed:
Quick fix

inner-daemons committed:
Removed unnecessary TODO statement

inner-daemons committed:
A

jimblandy commented on naga/src/valid/interface.rs:
```diff
@@ -628,6 +713,14 @@ impl super::Validator {
             }
         }
 
+        if var.space == crate::AddressSpace::TaskPayload {
+            let ty = &gctx.types[var.ty].inner;
+            // HLSL doesn't allow zero sized payloads.
+            if ty.try_size(gctx) == Some(0) {
+                return Err(GlobalVariableError::ZeroSizedTaskPayload);
+            }
+        }
```
In SPIR-V, variables in the `TaskPayloadWorkgroupExt` address class are not allowed to have initializers. Should Naga forbid initializers for `<task_payload>` globals?

inner-daemons replied:
The lines right below this have
```rust
if let Some(init) = var.init {
    match var.space {
        crate::AddressSpace::Private | crate::AddressSpace::Function => {}
        _ => {
            return Err(GlobalVariableError::InitializerNotAllowed(var.space));
        }
    }
    ...
}
```

In general, task payload address space is basically just workgroup address space with restrictions on which variables can be declared with it. Workgroup address space also can't have an initializer.

jimblandy replied:
You're totally right - I missed that. Thanks.

inner-daemons committed:
Tried to be more expressive

jimblandy committed:
validate that mesh shader entry ponits do have mesh info

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy commented on naga/src/valid/analyzer.rs:
```diff
@@ -1113,6 +1147,34 @@ impl FunctionInfo {
                 }
+                S::MeshFunction(func) => match &func {
+                    // TODO: double check all of this uniformity stuff. I frankly don't fully understand all of it.
+                    &crate::MeshFunction::SetMeshOutputs {
```
I don't see anything that would prevent this function from being called outside a mesh shader and its callees. That should be an error, right?

jimblandy reviewed (changes requested):
There should be a new `Capabilities` flag for mesh shading, and the validator should reject mesh stages unless it is set. If that's in a subsequent PR, it needs to be pulled forward into this one; we can't have the stuff in the IR unless we forbid it when it's not enabled. I know the backends will just crash; but it's easiest to keep things coherent.

Please remove the TODO comments and file a follow-up issue for each one if appropriate.

jimblandy reviewed (dismissed):
The entry point validation in `Validator::validate_module_handles` needs to handle `EntryPoint::mesh_info` and `EntryPoint::task_payload` as well.

cwfitzgerald commented:
> There should be a new Capabilities flag for mesh shading, and the validator should reject mesh stages unless it is set. If that's in a subsequent PR, it needs to be pulled forward into this one; we can't have the stuff in the IR unless we forbid it when it's not enabled. I know the backends will just crash; but it's easiest to keep things coherent.

@jimblandy we don't currently do this for compute shaders either though - _naga_ isn't the one who is validating the stage is valid, it's wgpu

jimblandy commented:
Compute shaders are a base WebGPU feature, though, whereas mesh shaders are not.

cwfitzgerald commented:
Oh I was thinking about _backends_ once we already know the stages involved, yes the validator needs to be validating out mesh entry points (and compute entry points too; but that doesn't have to be here)

inner-daemons committed:
Made functions only work in mesh shader entry points

inner-daemons committed:
Various validation fix attempts

inner-daemons commented:
I have changed `Capabilities` from being a `u32` to a `u64`, since we just reached 30/32 bits used. I can revert that if it doesn't belong in this PR for some reason.

cwfitzgerald commented:
We only need to bump it once we actually hit 33 :)

inner-daemons commented:
Ok, I'll revert that.

inner-daemons committed:
Undid capabilities resize

inner-daemons commented:
Just put up #8370, which is the WGSL parsing PR. The WGSL parsing and validation passes there, so this PR's validation seems to not be disallowing valid shaders yet.

inner-daemons committed:
Erich recommendations from #8370

inner-daemons committed:
Updated docs & validation for some builtins

inner-daemons committed:
Added some docs & removed contentious "// TODO"

inner-daemons commented:
@jimblandy Me and @cwfitzgerald agreed that mesh shader outputs should always be buffered, so I'm gonna remove references in the spec to having to call `setMeshOutputs` exactly once before calling `setVertex`/`setPrimitive` and stuff like that.

inner-daemons committed:
Spec docs changes

jimblandy commented:
> @jimblandy Me and @cwfitzgerald agreed that mesh shader outputs should always be buffered, so I'm gonna remove references in the spec to having to call `setMeshOutputs` exactly once before calling `setVertex`/`setPrimitive` and stuff like that.

Does this mean we could also potentially have a design where the vertex and primitive arrays and counts are returned by the mesh shader? It seemed to me like much of this stuff would work just as well as new `@builtin` kinds:

```wgsl
struct MeshOutput {
    @builtin(vertex_count) num_vertices: u32,
    @builtin(primitive_count) num_primitives: u32,
    @builtin(vertices) vertices: array<Vertex, 10>,
    @builtin(primitives) primitives: array<Primitive, 15>,
}

@mesh @workgroup_size(15, 1, 1)
fn mesh_shader(
    @builtin(local_invocation_index) invocation: u32,
    @builtin(task_payload) payload: Payload
) -> MeshOutput {
    ...
}
```

inner-daemons commented:
> > @jimblandy Me and @cwfitzgerald agreed that mesh shader outputs should always be buffered, so I'm gonna remove references in the spec to having to call `setMeshOutputs` exactly once before calling `setVertex`/`setPrimitive` and stuff like that.
> 
> 
> 
> Does this mean we could also potentially have a design where the vertex and primitive arrays and counts are returned by the mesh shader? It seemed to me like much of this stuff would work just as well as new `@builtin` kinds:
> 
> 
> 
> ```wgsl
> 
> struct MeshOutput {
> 
>     @builtin(vertex_count) num_vertices: u32,
> 
>     @builtin(primitive_count) num_primitives: u32,
> 
>     @builtin(vertices) vertices: array<Vertex, 10>,
> 
>     @builtin(primitives) primitives: array<Primitive, 15>,
> 
> }
> 
> 
> 
> @mesh @workgroup_size(15, 1, 1)
> 
> fn mesh_shader(
> 
>     @builtin(local_invocation_index) invocation: u32,
> 
>     @builtin(task_payload) payload: Payload
> 
> ) -> MeshOutput {
> 
>     ...
> 
> }
> 
> ```

Err, that's actually really clever. I do think that this isn't right though, since the thing output is per workgroup and this would have each thread return a giant structure, that would in theory be the same for all threads anyway. I'm not opposed to something closer to this though, like a global thread group variable that holds the buffered outputs and then the mesh shader can declare that it uses that global variable.

inner-daemons commented:
In fact, if we could have each of them in their own global variable, it would match closely to what vulkan has. So then if we wanted trusted shaders that followed stricter rules to avoid copies then we could do that.

I think if we wanted to enable that though, then setMeshOutputs at least would still have to be an intrinsic and we shouldn't replace it with a global variable.

I'll throw up a sketch. Thanks for thinking about this so thoroughly!

inner-daemons commented:
Ok, I've thought about it some more. I think that we can't really avoid buffering for another reason: the output format isn't guaranteed to match whatever the naga struct is. SPIR-V for example has arcane rules on which values must be in different output arrays. HLSL is very similar. I still don't think we should have the task payload/mesh outputs be per-thread, but a small tweak to your proposal would be as follows:

```rust
struct MeshOutput {
    @builtin(vertex_count) num_vertices: u32,
    @builtin(primitive_count) num_primitives: u32,
    @builtin(vertices) vertices: array<Vertex, 10>,
    @builtin(primitives) primitives: array<Primitive, 15>,
}

var<mesh_output> output: MeshOutput;
var<task_payload> task_payload: Payload;

@mesh(output) @workgroup_size(15, 1, 1)
fn mesh_shader(
    @builtin(local_invocation_index) invocation: u32,
) {
    ...
}
```

Then we would just need to allow these `@builtin` attributes on fields in a `mesh_output` variable. This variable would then probably function basically the same as `task_payload` does in task shaders, where only 1 can be used by a task shader and it must match what was declared, and other stages can't access such a global variable.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy committed:
check for task_payload globals without Capabilities::MESH_SHADER

jimblandy commented:
@inner-daemons Have the commits I've pushed been looking okay?

inner-daemons commented:
@jimblandy They look great! I very much appreciate them.

inner-daemons commented:
@jimblandy Do you have opinions about what we should go with? We could completely switch it up for the new option, which would probably mean less work on each of the backends. On the other hand, I would have to redo most of the work for mesh shaders, and there is a lot of work to redo.

I'm of the opinion that unless there's some problem with the new option I haven't seen we should go with that, but it would be kind of heartbreaking to restart after writing 7k lines and being blocked for 2 months on this PR.

cwfitzgerald commented:
Been meaning to sit down with this and figure it out, but ended up getting sick over the weekend ;(

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

jimblandy commented:
Writing down what we said in our meeting today:

- This PR is in no way blocked on the new suggestion for the API. Let's get this landed and iterate.
- It seems like the new suggestion might simplify SPIR-V generation.
- We should make sure our design doesn't hurt the performance of SPIR-V input.
- The WGSL front end work should wait until #8386 has landed.

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

inner-daemons committed:
Merge branch 'trunk' into mesh-shading/naga-ir

cwfitzgerald reviewed (approved):
I'm tired of looking at this, you're tired of looking at this, I don't see any immediate problems, lets land and iterate!

LukasKalbertodt commented:
Sorry for the spam, but I just want to say: thank you all for the work on this! I know this is not the first nor last step in the mesh shader journey, but as a usually quiet bystander I felt I should express my gratitude here :)
